### PR TITLE
Avoid saturating channels before we split payments

### PR DESCRIPTION
**Pull request by TheBlueMatt**
Currently we only opt to split a payment into an MPP if we have
completely and totally used a channel's available capacity (up to
the announced htlc_max or on-chain capacity, whichever is lower).
This is obviously destined to fail as channels are unlikely to have
their full capacity available.

Here we do the minimum viable fix by simply limiting channels to
only using up to a configurable power-of-1/2. We default this new
configuration knob to 1 (1/2 of the channel) so as to avoid a
substantial change but in the future we may consider changing this
to 2 (1/4) or even 3 (1/8).

In a (somewhat optional, though I'd prefer it) second commit, we ensure we don't spuriously fail to find a route just because of the saturation limit, opting to remove the saturation limit during pathfinding if we stop finding new paths. This at least notably means we don't have to have a ton of test changes, as it makes the behavior almost entirely backwards compatible even across various edge cases.

### CODE REVIEW

File: lightning/src/routing/router.rs [hunk 1]
```diff
@@ -233,6 +243,7 @@ impl_writeable_tlv_based!(PaymentParameters, {
 	(2, features, option),
 	(3, max_path_count, (default_value, DEFAULT_MAX_PATH_COUNT)),
 	(4, route_hints, vec_type),
+	(5, max_channel_saturation_power_of_half, (default_value, 1)),
```
[the remark this answers is not in the document]
Comment (dunxen) [reply lightning/src/routing/router.rs:250]: Ah of course, part of the "okay to be odd" rule.

**Comment by moneyball** [comment]
Concept ACK

File: lightning/src/routing/router.rs [hunk 2]
```diff
@@ -1505,12 +1552,14 @@ where L::Target: Logger {
 						.entry((hop.candidate.short_channel_id(), *prev_hop < hop.node_id))
 						.and_modify(|used_liquidity_msat| *used_liquidity_msat += spent_on_hop_msat)
 						.or_insert(spent_on_hop_msat);
-					if *used_liquidity_msat == hop.candidate.htlc_maximum_msat() {
+					let hop_capacity = hop.candidate.effective_capacity();
+					let hop_max_msat = max_htlc_from_capacity(hop_capacity, channel_saturation_pow_half);
```
Comment (tnull) [inline lightning/src/routing/router.rs:1556]: nit: `s/hop_max_msat/hop_htlc_max_msat/`

File: lightning/src/routing/router.rs [hunk 3]
```diff
@@ -225,6 +225,16 @@ pub struct PaymentParameters {
...
+
+	/// Selects the maximum percent of a channel's total capacity which will be sent over a
+	/// channel, as a power of 1/2. A higher value prefers to send the payment using more MPP parts
+	/// whereas a lower value prefers to send larger MPP parts, potentially saturating channels.
+	///
+	/// A value of 0 will allow payments up to and including a channel's total announced usable
+	/// capacity, a value of one will only use up to half its capacity, two 1/4, etc.
+	///
+	/// Default value: 1
+	pub max_channel_saturation_power_of_half: u8,
```
[the remark this answers is not in the document]
Comment (tnull) [reply lightning/src/routing/router.rs:241]: Fair enough, I assumed this was the main reason for it. While there might be ways to work around the floating point division, your solution is probably cleaner, even though it has the drawback that no saturation between 0.5 and 1.0 times the capacity can be specified.

File: lightning/src/routing/router.rs [hunk 4]
```diff
@@ -225,6 +225,20 @@ pub struct PaymentParameters {
 	/// The maximum number of paths that may be used by (MPP) payments.
 	/// Defaults to [`DEFAULT_MAX_PATH_COUNT`].
 	pub max_path_count: u8,
+
+	/// Selects the maximum percent of a channel's total capacity which will be sent over a
```
Comment (tnull) [inline lightning/src/routing/router.rs:229]:
```suggestion
	/// Selects the maximum share of a channel's total capacity which will be sent over a
```

File: lightning/src/routing/router.rs [hunk 5]
```diff
@@ -225,6 +225,20 @@ pub struct PaymentParameters {
...
+	/// channel, as a power of 1/2. A higher value prefers to send the payment using more MPP parts
+	/// whereas a lower value prefers to send larger MPP parts, potentially saturating channels.
```
Comment (tnull) [inline lightning/src/routing/router.rs:231]:
```suggestion
	/// whereas a lower value prefers to send larger MPP parts, potentially saturating channels and increasing the failure probability.
```

**Review by tnull — APPROVED** [review]
LGTM

**Review by tnull — DISMISSED** [review]

**Review by tnull — APPROVED** [review]
